### PR TITLE
Implement imperative infer var type

### PR DESCRIPTION
**Pull request by velconia**
1. InferVarType for both imperative and static mode

### CODE REVIEW

File: paddle/fluid/framework/var_type_inference.h [hunk 1]
```diff
@@ -14,33 +14,132 @@ limitations under the License. */
...
+
+  virtual const std::vector<std::string>& Output(
+      const std::string& name) const {
+    PADDLE_ENFORCE_NOT_NULL(op_);
+    return op_->Output(name);
+  }
+
+  virtual proto::VarType::Type GetType(const std::string& name) const {
+    PADDLE_ENFORCE_NOT_NULL(block_);
+    return block_->FindRecursiveOrCreateVar(name).GetType();
```
Comment (panyx0718) [inline paddle/fluid/framework/var_type_inference.h:69]: Better not to create. Does test fail if you don't create?

`MaybeCreateVarAndGetType`?

File: paddle/fluid/framework/var_type_inference.h [hunk 2]
```diff
@@ -14,33 +14,132 @@ limitations under the License. */
...
+  }
+
+  virtual void SetType(const std::string& name, proto::VarType::Type type) {
+    PADDLE_ENFORCE_NOT_NULL(block_);
+    block_->FindRecursiveOrCreateVar(name).SetType(type);
```
Comment (panyx0718) [inline paddle/fluid/framework/var_type_inference.h:74]: same as above, same below

File: paddle/fluid/imperative/layer.h [hunk 3]
```diff
@@ -238,7 +258,7 @@ class VarBase {
   }
 
   std::string name_;
-  framework::proto::VarType::Type dtype_;
+  framework::proto::VarType::Type type_;
   platform::Place place_;
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.h:262]: still used?

File: paddle/fluid/imperative/layer.h [hunk 4]
```diff
@@ -359,12 +381,131 @@ class PyLayer {
...
+  }
+
+  void SetDataType(const std::string& name,
+                   framework::proto::VarType::Type type) override {
+    var_set_[name]->SetDataType(type);
+  }
+
+  std::vector<framework::proto::VarType::Type> GetDataTypes(
+      const std::string& name) const override {
+    PADDLE_THROW("GetDataTypes is not supported in runtime InferVarType");
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.h:476]: why?

File: paddle/fluid/imperative/tracer.cc [hunk 5]
```diff
@@ -160,7 +161,7 @@ Tracer::Tracer(framework::BlockDesc* root_block) : root_block_(root_block) {
 }
 
 std::set<std::string> Tracer::Trace(OpBase* op, const VarBasePtrMap& inputs,
-                                    const VarBasePtrMap& outputs,
+                                    VarBasePtrMap& outputs,
```
Comment (panyx0718) [inline paddle/fluid/imperative/tracer.cc]: pointer?

File: paddle/fluid/pybind/imperative.cc [hunk 6]
```diff
@@ -48,7 +48,7 @@ void BindTracer(pybind11::module* m) {
       .def("trace",
            [](imperative::Tracer& self, imperative::OpBase* op,
               const imperative::VarBasePtrMap& inputs,
-              const imperative::VarBasePtrMap& outputs,
+              imperative::VarBasePtrMap& outputs,
```
Comment (panyx0718) [inline paddle/fluid/pybind/imperative.cc]: pointer?

File: paddle/fluid/framework/details/graph_test_base.h [hunk 7]
```diff
@@ -68,11 +68,11 @@ class SplitOpMaker : public OpProtoAndCheckerMaker {
 
 class DummyVarTypeInference : public VarTypeInference {
  public:
-  void operator()(const OpDesc& op_desc, BlockDesc* block) const override {
-    auto& inputs = op_desc.Input("X");
-    auto type = block->Var(inputs.front())->GetType();
-    auto out_var_name = op_desc.Output("Out").front();
-    block->Var(out_var_name)->SetType(type);
+  void operator()(framework::InferVarTypeContext& ctx) const override {
```
Comment (panyx0718) [inline paddle/fluid/framework/details/graph_test_base.h]: const? or pointer?

File: paddle/fluid/framework/var_type_inference.h (hunk 1 above)
Comment (velconia) [reply paddle/fluid/framework/var_type_inference.h:69]: this is copied from the original code, just keep the same as much as possible I guess

File: paddle/fluid/imperative/layer.h (hunk 3 above)
Comment (velconia) [reply paddle/fluid/imperative/layer.h:262]: yes

File: paddle/fluid/imperative/tracer.cc (hunk 5 above)
Comment (velconia) [reply paddle/fluid/imperative/tracer.cc]: Done

File: paddle/fluid/pybind/imperative.cc (hunk 6 above)
Comment (velconia) [reply paddle/fluid/pybind/imperative.cc]: Done

File: paddle/fluid/framework/details/graph_test_base.h (hunk 7 above)
Comment (velconia) [reply paddle/fluid/framework/details/graph_test_base.h]: Done

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**